### ISSUE TITLE
Missing synchronizations after Hiding semaphores from the compiler

### ISSUE THREAD

**Issue by raikonenfnu** [openxla/iree#10483]
After this commit https://github.com/iree-org/iree/commit/9ef8abfee1c25e10fdc026db26c9f72662c3c4fe, our LevelZero backend seemed to break by always returning zeroes. I think we use to have similar issue until we added execution barrier https://github.com/raikonenfnu/iree/blob/ze-shark-09-19-2022/experimental/level_zero/direct_command_buffer.c#L158-L174. I assume I am missing some implementation for fences/semaphore somewhere.

**Comment by benvanik** [comment]
Make sure your backend adds the legacy_sync attr: https://github.com/google/iree/blob/main/compiler/src/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp#L129-L131

**Comment by raikonenfnu** [comment]
ahh I added it, however the error still seem to persist 😅

**Comment by benvanik** [comment]
there's no known issues with barriers within command buffers - vulkan and cpu both take advantage of them - so I'd check your usage of the API or try to build a reproducer?

**Comment by raikonenfnu** [comment]
@benvanik I was wondering if you can point to where the fences are being called in vulkan, we may be missing some implementation for our [LevelZero HAL](https://github.com/raikonenfnu/iree/tree/ze-shark-09-19-2022/experimental/level_zero), is [this](https://github.com/raikonenfnu/iree/blob/ze-shark-09-19-2022/experimental/level_zero/direct_command_buffer.c#L158-L174) still the right place?

**Comment by benvanik** [comment]
That's for the inside of command buffers only and only execution barriers - the device queue operations are where fences/semaphores are - ensure your device.c file doesn't have "// TODO: implement" in the queue methods and that wait/signal semaphore lists are handled (iree_hal_semaphore_list_wait on waits, iree_hal_semaphore_list_signal on signals).

**Comment by raikonenfnu** [comment]
@benvanik unfortunately I was able to find iree_hal_semaphore_list_wait on waits, iree_hal_semaphore_list_signal on the queue method. I was wondering if after this change(hiding semaphore), we would need the a device-based semaphore implemented since, since right now it's following cuda's semaphore which seems to be basically from host and using IREE apis such as `iree_hal_semaphore_poll` and `iree_atomic_load_int64`?

**Comment by raikonenfnu** [comment]
would you also happen to know a good pass to check the dumps/ir after to compare the difference between hiding and no hiding semaphores? :)

**Comment by benvanik** [comment]
There should have been no material changes from that change in a correctly operating backend in legacy_sync mode so I'm not sure looking at IR will help much, but if you look before/after createFixupLegacySyncPass (or the canonicalize after it) you'll see the intended execution ordering and the actual one you should see at runtime. Effectively the only difference at runtime is that instead of a single shared semaphore that has its value incremented repeatedly now there's a unique semaphore per submission and it only transitions 0->1.

CUDA's semaphores were never fully implemented - poll and atomic ints are not the right way to implement them and only work because of the "TODO: remove me" host/device synchronization calls that are sprinkled around. I've still got a little reworking to do to simplify HAL synchronization (possibly removing semaphores entirely and just treating fences as events) but what's in the CUDA impl today should be sufficient with legacy_sync. I honestly have no idea how many assumptions the entirety of the backend is making about it not running asynchronously and how to manage that. I expect nearly all of it to be rewritten when it actually migrates to async (and the legacy_sync flag is dropped).

Your next step is to add printfs or verify your backend matches CUDA with respect to all of the places it synchronizes including those implicit to the CUDA API that are being relied upon. That the CUDA backend wasn't finished and made async to the old way of doing things means it's currently held together by implicit assumptions founded on TODOs and if any of your semantics differ it's unlikely to work.

**Comment by raikonenfnu** [comment]
I see, thanks for the explanation Ben. Also fwiw, to expand upon your mention of semantic and assumptions difference between L0 and cuda, we needed to add the execution_barrier manually for L0. 

I was wondering for
```
Your next step is to add printfs or verify your backend matches CUDA with respect to all of the places it synchronizes including those implicit to the CUDA API that are being relied upon
```
do I just print in execution_barrier and queue_execute and ensure they give the same print order and quantity?

**Comment by raikonenfnu** [comment]
Hey @benvanik printing out the functions in semaphores.c, it seems to confirm what you mentioned earlier about creating a new semaphore and waiting on that one instead of reusing/sharing the same one.
Before hiding semaphores:
```bash
creating 0x560025dfa5c0: 0
query 0x560025dfa5c0: 0
query 0x560025dfa5c0: 0
wait 0x560025dfa5c0: 0
EXEC @forward
query 0x560025dfa5c0: 0
query 0x560025dfa5c0: 0
wait 0x560025dfa5c0: 0
signalling 0x560025dfa5c0: 2
query 0x560025dfa5c0: 2
query 0x560025dfa5c0: 2
query 0x560025dfa5c0: 2
query 0x560025dfa5c0: 2
query 0x560025dfa5c0: 2
wait 0x560025dfa5c0: 2
query 0x560025dfa5c0: 2
query 0x560025dfa5c0: 2
wait 0x560025dfa5c0: 2
query 0x560025dfa5c0: 2
query 0x560025dfa5c0: 2
wait 0x560025dfa5c0: 2
``` 

After hiding semaphores:
```bash
creating 0x55dc3243f0d0: 0
query 0x55dc3243f0d0: 0
query 0x55dc3243f0d0: 0
wait 0x55dc3243f0d0: 0
EXEC @forward
creating 0x55dc32b766b0: 0
query 0x55dc3243f0d0: 0
query 0x55dc3243f0d0: 0
wait 0x55dc3243f0d0: 0
signalling 0x55dc32b766b0: 1
query 0x55dc32b766b0: 1
query 0x55dc32b766b0: 1
creating 0x55dc32af1e40: 0
query 0x55dc32b766b0: 1
query 0x55dc32af1e40: 0
query 0x55dc32af1e40: 0
wait 0x55dc32af1e40: 0
creating 0x55dc328a85d0: 0
query 0x55dc32af1e40: 0
query 0x55dc32af1e40: 0
wait 0x55dc32af1e40: 0
query 0x55dc328a85d0: 0
query 0x55dc328a85d0: 0
wait 0x55dc328a85d0: 0
```

Although I was wondering if it is expected for one of the semaphore that was signalled to not have anything waiting on it?
Or if do you think there is something that looks ridiculously wrong? Thanks in advance for the guidance. 🙂

**Comment by raikonenfnu** [comment]
new update:

If we force the code to use the same semaphores every time we run `iree_hal_level_zero_device_create_semaphore`:
```diff
@@ -293,8 +294,12 @@ static iree_status_t iree_hal_level_zero_device_create_semaphore(
     iree_hal_semaphore_t** out_semaphore) {
   iree_hal_level_zero_device_t* device =
       iree_hal_level_zero_device_cast(base_device);
-  return iree_hal_level_zero_semaphore_create(&device->context_wrapper,
-                                              initial_value, out_semaphore);
+  if(device->semaphore == NULL) {
+    iree_hal_level_zero_semaphore_create(&device->context_wrapper,
+                                                initial_value, &device->semaphore);
+  }
+  *out_semaphore = device->semaphore;
+  return iree_ok_status();
 }
 ```
 It works 🤔

**Comment by benvanik** [comment]
That won't work for long as it's not retaining the semaphore it's returning - but more importantly you can't share semaphores like that as that would cause timelines to go backwards.

**Comment by raikonenfnu** [comment]
> That won't work for long as it's not retaining the semaphore it's returning - but more importantly you can't share semaphores like that as that would cause timelines to go backwards.

Haha, I know, was just an observation. :)

**Comment by nithinsubbiah** [comment]
Hi @benvanik, I'm looking into this issue again.

tldr: We added a Level Zero HAL backend in IREE to target Intel GPU's which is giving incorrect results after this patch. If I revert this commit, it gives the right results.

Incorrect results are obtained for MLIR files which loads values ex: weights, bias as in https://github.com/iree-org/iree/blob/main/samples/models/mnist.mlir#L15. If I pass those values as arguments then the results match.

I ran the semaphore cts tests for the backend which passes the same number of tests with and without this patch.  

It would be really helpful if you have any pointers on what could cause this regression, thanks! :)

**Comment by benvanik** [comment]
Unfortunately the CTS is woefully incomplete ;(
Sounds like a bug in the level zero backend. I can take a look at a trace from tracy if you grab one, but I suspect it's an issue of the implementation not waiting when it should and something you'll need to debug through. The trace will help reveal the intended flow of the program and make it easier to see where things may be getting out of order.

**Comment by ScottTodd** [comment]
(After more debugging) If you can narrow down a test case and expand the CTS that would be great. What's in there today was written after a few reads through the interface defined in the header file and some observations when running with the in-tree HAL backends. Though that being said... testing for synchronization is generally much harder than testing synchronous code.

**Comment by nithinsubbiah** [comment]
Sure Ben, I'll get a tracy profile of a program that gives wrong results on the backend. 

@ScottTodd Will definitely look into expanding the CTS covering this bug once I narrow down this issue.

**Comment by nithinsubbiah** [comment]
@benvanik Here's the trace of running mnist model on Level Zero backend:
[mnist_levelzero.tracy.zip](https://github.com/iree-org/iree/files/10562384/mnist_levelzero.tracy.zip)

**Comment by benvanik** [comment]
Can't spot too much there with sampling as there's no debug symbols in your runtime build, and there are some big performance wtfs (like iree_hal_level_zero_direct_command_buffer_create averaging 300-400us and context switching half a dozen times on average), but given that you are synchronizing everything after every submission there's no reason you shouldn't be getting correct results unless there's some other issue in the level zero HAL or code generation. Sorry, this is probably going to entail someone familiar with that code to become more familiar with that code - if the program works on --device=vulkan and --device=local-task but not on --device=level-zero (or w/e) then it's something in the level zero implementation.

**Comment by nithinsubbiah** [comment]
I see okay, thanks for taking a look at this.

**Comment by allieculp** [comment]
@nithinsubbiah Would you like to leave this issue open, or can we close?

**Comment by nithinsubbiah** [comment]
We can close this issue, the problem was with the LevelZero API
